### ISSUE TITLE
List.append ignoring lifetime constraints when its eltType is a class

### ISSUE THREAD

**Issue by dlongnecke-cray** [chapel-lang/chapel#13613]
In https://github.com/chapel-lang/chapel/pull/13530, I added a new test based on the test `test/classes/delete-free/lifetimes/bad-push-back-borrow.chpl` to cover the `list` type.

The original test fails with a lifetime error, but the new test compiles without incident.

Code in question:

```chapel
pragma "safe"
module BadAppendBorrow {

use Lists;

class I {
  var dom: domain(1);
  var r: [dom] string;
  proc init(in r: [?dom] string) {
    this.dom = dom;
    this.r = r;
  }
}

class D {
  var iArr: list(borrowed I);
}

var i = new shared I(['a', 'b']);
var d = new owned D();

proc addone()
{
  // This should result in lifetime checking errror.
  //  * shared I destroyed at the end of this function
  //  * a borrow of it is saved in d.iArr.
  d.iArr.append(new shared I(i.r));
}

addone();

for element in d.iArr {
  writeln(element);
}

}
```

This test compiles silently even when the signature of `list.append` is the following (may need to update the code of append if it's not like this on master):

```
    /*
      Add an element to the end of this list.

      :arg x: An element to append.
    */
    proc append(pragma "no auto destroy" in x: eltType) lifetime this < x { ... }
```

**Comment by dlongnecke-cray** [comment]
Just a small update, the following program:

```chapel

private use Lists;

class Foo {
  var x: int = 0;
}

record someRecord {
  type eltType;
  var x: int = 0;

  proc init(type eltType, x: int=0) {
    this.eltType = eltType;
    this.x = x;
  }
}

var a: [1..0] borrowed Foo;
var b: list(borrowed Foo);
var c: someRecord(borrowed Foo);

writeln(a.type:string);
writeln(a.eltType == borrowed Foo);


writeln(b.type:string);
writeln(b.eltType == borrowed Foo);

writeln(c.type:string);
writeln(c.eltType == borrowed Foo);
```

Gives the following output:

```
[domain(1,int(64),false)] borrowed Foo
true
list(Foo,false)
true
someRecord(Foo)
true
```

So I guess the problem must lie elsewhere (and a separate issue made for record types not properly printing out class decorators).

**Comment by mppf** [comment]
@dlongnecke-cray - just wondering if you have fixed this already or are working on it? I seem to remember a related recent discussion.

**Comment by dlongnecke-cray** [comment]
Yes, this was fixed by #15575.
